records/2024-02-15.private-eks-cluster.md

```diff
@@ -0,0 +1,94 @@
+# Converting To A Private EKS Cluster
+
+Date: 2024-02-15
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+GC Notify hosts a large portion of its system in an AWS managed Elastic Kubernetes Service. When deploying an EKS cluster, there is the option to make the main administrative endpoint either public or private. Public endpoints are available to the entirety of the internet, but protected by the AWS IAM and Single-Sign-On. Private endpoints are also protected by AWS IAM and SSO but are also not accessible via the open internet. 
+
+Currently, GC Notify is using a public administration endpoint, which is typically not recommended for sensitive workloads. This ADR will go over the pros and cons of public vs private EKS clusters.
+
+## Options
+
+### Public EKS Cluster
+
+Using a public EKS cluster means keeping that status-quo for GC Notify. When the EKS cluster is deployed, a random AWS DNS entry is created, which resolves to a public IP exposing the GC Notify EKS API endpoint.  
+
+```mermaid
+graph TD;
+    GitHub-->Internet;
+    Administrator-->Internet;
+    Internet-->EKS;
+```
+#### Considerations
+
+##### Access Control
+
+Access control is provided by using the CDS AWS SSO solution. A user that has authenticated with AWS and has the necessary permissions is able to log in and administer the cluster.
+
+##### Security
+
+Because the API is exposed as a public endpoint, a user would be able to access the API from any system anywhere in the world, so long as they are able to authenticate with CDS and AWS. This provides flexibility for supporting the system.
+
+This also means that the API is exposed to malicious actors. Although authentication is required to actually access the system, this still leaves it vulnerable to zero day attacks, DDoS attacks, and other security threats. 
+
+This may not conform to AC-17 of ITSG-33 government security requirements.
+
+##### Integration With GitHub
+
+Being a public endpoint, it is possible to provide GitHub direct connectivity to the cluster API. This has the benefit of simplifying release management with our CICD workflows in GitHub Actions. By granting GitHub access, any Kubernetes or Helm changes can be executed from the public GitHub Action Runners.
+
+##### Administrative Overhead
+
+There is little to no administrative overhead when using a public endpoint since this is the default state for AWS EKS clusters. 
+
+### Private EKS Cluster
+
+Using a private EKS Cluster requires the use of a VPN to connect to the cluster for administrative tasks. When the cluster is deployed, it will be unreachable unless connected via our AWS VPN endpoint for that particular environment.
+
+
+```mermaid
+graph TD;
+    EKS-->Internet-->GitHub;
+    Administrator-->VPN-->EKS;
+```
```

Comment:
It might be beneficial to have a workflow diagram with a bit more details for this one. We can quickly sync offline about it.